### PR TITLE
Custom procedure remove button

### LINKED ISSUE #1241: Procedures: delete icon
The delete icon is a singleton--it's always associated with the selected argument or label.

image

Who owns it: the field being edited?  The procedure declaration editor?

When is it visible?  When a component is selected *and* there are multiple components on the block.

How does it know which component it's associated with?  It needs that to delete the correct component.

### PR DESCRIPTION
**Pull request by paulkaplan**
### Resolves

_What Github issue does this resolve (please include link)?_
Fixes https://github.com/LLK/scratch-blocks/issues/1241
### Proposed Changes

_Describe what this Pull Request does_
One commit on top of the other custom procedures work that hasn't landed yet, so only look at the last commit https://github.com/LLK/scratch-blocks/commit/49ad421605001258e499dfd2dcbad00b5652c803

I wasn't sure how to chain together the click action for the remove button with the actual declaration removal code, so that is definitely something I want feedback on @rachel-fenichel.

### CODE REVIEW

File: blocks_vertical/procedures.js [hunk 1]
```diff
@@ -565,6 +565,53 @@ Blockly.ScratchBlocks.ProcedureUtils.addStringNumberExternal = function() {
...
+      if (target.getField(field.name) == field) {
+        Blockly.WidgetDiv.hide(true);
+        this.removeInput(name);
+        this.onChangeFn();
+        this.updateDisplay_();
+      }
+    } else {
+      for (var j = 0; j < input.fieldRow.length; j++) {
+        if (input.fieldRow[j] == field) {
+          Blockly.WidgetDiv.hide(true);
```
Comment (rachel-fenichel) [inline blocks_vertical/procedures.js]: You're doing the same things here and in lines 585-588.  Instead have a variable `nameToRemove` declared outside of the loop and set to null.  In the loop, if the checks pass, set `nameToRemove` to `name` and break.  Then do all of these steps based on `nameToRemove`, if it's not null.

File: core/block.js [hunk 2]
```diff
@@ -1477,6 +1480,23 @@ Blockly.Block.newFieldTextInputFromJson_ = function(options) {
   return field;
 };
 
+/**
```
Comment (rachel-fenichel) [inline core/block.js]: Move this into your new field file.  I just filed https://github.com/google/blockly/issues/1486 with an explanation + plans to do the same for the default field types.

File: core/css.js [hunk 3]
```diff
@@ -163,7 +163,7 @@ Blockly.Css.CONTENT = [
   '}',
 
   '.blocklyWidgetDiv.fieldTextInput {',
-    'overflow: hidden;',
+    // 'overflow: hidden;',
```
Comment (rachel-fenichel) [inline core/css.js]: ?

File: core/field_textinput_removable.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2012 Google Inc.
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js]: Update license year, fileoverview, and author.

File: core/field_textinput_removable.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * @fileoverview Text input field.
+ * @author fraser@google.com (Neil Fraser)
+ */
+'use strict';
+
+goog.provide('Blockly.FieldTextInputRemovable');
+
+goog.require('Blockly.FieldTextInput');
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js]: Alphabetize the Blockly.* requires, per https://google.github.io/styleguide/jsguide.html#file-goog-require

File: core/field_textinput_removable.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+goog.require('Blockly.Msg');
+goog.require('Blockly.utils');
+goog.require('goog.asserts');
+goog.require('goog.dom');
+goog.require('goog.dom.TagName');
+goog.require('goog.userAgent');
+
+
+/**
+ * Class for an editable text field.
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js]: Class for an editable text field which displays a deletion icon when selected.

File: core/field_textinput_removable.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+ * @param {string} text The initial content of the field.
+ * @param {Function=} opt_validator An optional function that is called
+ *     to validate any constraints on what the user entered.  Takes the new
+ *     text as an argument and returns either the accepted text, a replacement
+ *     text, or null to abort the change.
+ * @param {RegExp=} opt_restrictor An optional regular expression to restrict
+ *     typed text to. Text that doesn't match the restrictor will never show
+ *     in the text field.
+ * @extends {Blockly.Field}
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js]: @extends {Blockly.FieldTextInput}

File: core/field_textinput_removable.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ * @private
+ */
+Blockly.FieldTextInputRemovable.prototype.showEditor_ = function() {
+  Blockly.FieldTextInputRemovable.superClass_.showEditor_.call(this);
+
+  var div = Blockly.WidgetDiv.DIV;
+  var removeButton =
+      goog.dom.createDom(goog.dom.TagName.IMG, 'blocklyTextDropDownArrow');
+  removeButton.setAttribute('src',
+    Blockly.mainWorkspace.options.pathToMedia + 'icons/remove.svg');
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js]: Indent +2 spaces

File: core/field_textinput_removable.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * Show the inline free-text editor on top of the text with the remove button.
+ * @private
+ */
+Blockly.FieldTextInputRemovable.prototype.showEditor_ = function() {
+  Blockly.FieldTextInputRemovable.superClass_.showEditor_.call(this);
+
+  var div = Blockly.WidgetDiv.DIV;
+  var removeButton =
+      goog.dom.createDom(goog.dom.TagName.IMG, 'blocklyTextDropDownArrow');
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js]: `blocklyTextDropdownArrow`?

File: core/field_textinput_removable.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  removeButton.setAttribute('src',
+    Blockly.mainWorkspace.options.pathToMedia + 'icons/remove.svg');
+  removeButton.style.width = 25 + 'px';
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js]: These are all constants.  Can you move them into CSS?

File: core/field_textinput_removable.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  removeButton.style['margin-left'] = -15 + 'px';
+
+  this.removeButtonMouseWrapper_ = Blockly.bindEvent_(removeButton,
+    'mousedown', this, this.removeCallback_);
+  div.appendChild(removeButton);
+};
+
+Blockly.FieldTextInput.prototype.removeCallback_ = function() {
+  if (this.sourceBlock_ && this.sourceBlock_.removeFieldCallback) {
+    this.sourceBlock_.removeFieldCallback(this);
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js]: This looks reasonable.  Add an else that warns the developer that they should have implemented removeFieldCallback on the source block, and of course add documentation.

File: core/block.js (hunk 2 above)
Comment (paulkaplan) [reply core/block.js]: If I do that, does this file need to be updated to actually use that function? Or is that done automatically somehow?

File: core/css.js (hunk 3 above)
Comment (paulkaplan) [reply core/css.js]: Woops fixed with custom class

File: core/block.js (hunk 2 above)
Comment (rachel-fenichel) [reply core/block.js]: You would change line 1346 (which you just added) to call the new function.

File: blocks_vertical/procedures.js [hunk 12]
```diff
@@ -575,29 +575,28 @@ Blockly.ScratchBlocks.ProcedureUtils.removeFieldCallback = function(field) {
   if (this.inputList.length === 1) {
     return;
   }
-
+  var inputNameToRemove = null;
   for(var n = 0; n < this.inputList.length; n++) {
```
Comment (rachel-fenichel) [inline blocks_vertical/procedures.js]: nit: space between `for` and `(`

**Review by rachel-fenichel — COMMENTED** [review]
This differs from the scratch 2.0 behaviour in that the remove icon is always shown, even if you can't actually remove the input.  Scratch 2.0 hides the icon if it's not usable.  Are you planning to implement that?

**Comment by paulkaplan** [comment]
Thanks for all the feedback. I updated with more commits fixing the comments. 

WRT the remove icon always showing, I will return to that in another pass. I'll file an issue.

File: core/field_textinput_removable.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+/**
+ * Show the inline free-text editor on top of the text with the remove button.
+ * @private
+ */
+Blockly.FieldTextInputRemovable.prototype.showEditor_ = function() {
+  Blockly.FieldTextInputRemovable.superClass_.showEditor_.call(this);
+
+  var div = Blockly.WidgetDiv.DIV;
+  div.className += ' removableTextInput';
```
Comment (rachel-fenichel) [inline core/field_textinput_removable.js:64]: Does this get removed at the correct time?  Or does the widget div continue to have this forever?

**Review by rachel-fenichel — APPROVED** [review]
One question, and LGTM

File: core/field_textinput_removable.js (hunk 13 above)
Comment (paulkaplan) [reply core/field_textinput_removable.js:64]: It gets removed correctly! Followed the pattern from the fieldtextinput which does a similar thing. It gets removed when the widget is hidden.

**Pull request merged**